File: Python3/sorting_algs/insert_sort.py
from typing import List

def insert_sort(arr: List[int]) -> List[int]:
    for j in range(1, len(arr)):                # starting from second element since a single element list is trivially sorted
        print("key: " + str(arr[j]))
        key = arr[j]                            # keeping track of current element 
        i = j - 1                               # index of prev element 
        while key < arr[i] and i >= 0:
             arr[i+1] = arr[i] 
             i -= 1
        arr[i+1] = key                          # every num till the index of the key is sorted 
        print("loop invariance in action: " + str(arr) + " till index " + str(j)) 

    return arr 


if __name__ == "__main__":     
    print(insert_sort([5,4,3,2,1]))

            

File: C++/sorting_algs/insert_sort.cpp
#include <iostream>

void insert_sort(int input_arr[], int arr_len) 
{    
    for (int j = 1; j < arr_len; j++) 
    {
        // current value is remembered
        int key = input_arr[j];     
        int i = j - 1; 
        
        // swap prev value with current value in loop for as long as it is greater
        while(key < input_arr[i] && i >= 0)
        {
            input_arr[i + 1] = input_arr[i];
            i--;   
        } 

        // finally put the key value in the array where it is greater than the num to its left
        input_arr[i + 1] = key; 
    }
}

int main()
{
    // example array
    int arr[] = {2,1,0,5,10,11,9,3,4};
    int arr_len = sizeof(arr) / sizeof(int); 

    insert_sort(arr, arr_len);  

    bool first = true; 

     for (int i : arr)
    {
        std::cout << (first ? "Sorted Array: " : ", ") << i;
        first = false;
    }

    return 0; 
}

